doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,223 @@
+
+A jinja2 template is written in a domain specific language, but inherited many
+consepts from python, which means that anyone familiar with python should
+learn to write jinja2 templates with ease. 
+
+Data object
+===========
+
+All templates have access to a data object called ``data``. Through this data
+object will the template have access to information like all the elements in
+the .config file, 
+
+.. code-block:: c
+
+	{{ data['config']['CONFIG_BOARD'] }}
+
+devicetree data,
+ 
+ .. code-block:: c
+
+	{% for device in data['devicetree']['devices'] %}
+	    {{ device }}
+	{% endfor %}
+
+
+and runtime information like defines that has been injected using the CODEGEN_DEFINES
+key word in the CMakeLists.txt file.
+
+.. code-block:: c
+
+	{{ data['runtime']['defines']['MY_DEFINE'] }}
+
+Filters
+=======
+
+Filters are functions that can be applied to data in jinja2 templates and are used to
+modify the data in one form or the other. The filters are applied to the data by using
+the `pipe` character 
+
+.. code-block:: c
+
+    {{ "Hello %s\\n"\|format("World!") }}
+
+
+Built-in
+--------
+
+Jinja2 has a set of built in filter that in many cases will be familiar to anyone who
+have written code in python. information about built in filter can be found here
+`Built-in Filters <http://jinja.pocoo.org/docs/2.10/templates/#list-of-builtin-filters>`_.
+
+.. code-block:: c
+
+        {{ "fOOBAR!"|capitalize() }} => Foobar!
+
+Custom
+------
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
+the jinja2 engine to tailer it for any particular application.
+
+There is no square root filter built into jinja, this can be added by using 
+the ``sqrt()`` function in python to create a jinja2 filter
+
+Python code:
+
+.. code-block:: python
+
+        def squarerootfilter(value):
+                return sqrt(value)
+
+        environment.filters['sqrt'] = squarerootfilter
+
+Jinja2 template:
+
+.. code-block:: c
+        
+        Square root of 5: {{ 5|sqrt() }}
+
+
+
+
+Extending and reusing templates
+===============================
+
+Templates can extend and/or reuse each other through multiple mechanisms. To be able
+extend or reuse a template, the template must be stored in one of the folders listed
+in the search path list. Search paths can be set using the SEARCH_PATH key word in the
+CMakeLists.txt file.
+
+Include
+-------
+
+Include works more or less like include pre-processor directive in C, and will
+insert the content in another template starting from the location of the include keyword.
+
+.. code-block:: c
+
+	{% include 'another_template.jinja2' %}
+
+
+Import
+------
+
+The import keyword will import all macros defined in another template into a namespace.
+Similar to how import in python works. 
+
+.. code-block:: c
+
+	{% import 'template_with_macros.jinja2' as macros %}
+
+	{{ macros.an_imported_macro() }}
```

Comment:
Do we need to document how Jinja2 works? Isn't that done best by Jinja2?

Note that we do not document how yaml, DeviceTree, Kconfig, C, or the CMake language behaves.

I understand we might be doing so because Codegen did so, but with Jinja2 this is different because Codegen forked the language and was forced to also fork the documentation.

Reply:
Well, it's already there so let's keep it for now, at least through the RFC process.